services/src/kilda-pce/src/main/java/org/openkilda/pce/impl/AvailableNetwork.java

```diff
@@ -0,0 +1,102 @@
+/* Copyright 2018 Telstra Open Source
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.pce.impl;
+
+import org.openkilda.model.Isl;
+import org.openkilda.model.Switch;
+import org.openkilda.model.SwitchId;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Semantically, this class represents an "available network". That means everything in it is available for path
+ * allocation.
+ */
+@Slf4j
+@ToString
+public class AvailableNetwork {
+    @VisibleForTesting
+    final Map<SwitchId, Switch> switches = new HashMap<>();
+
+    public Switch getSwitch(SwitchId dpid) {
+        return switches.get(dpid);
+    }
+
+    /**
+     * Creates switches (if they are not created yet) and ISL between them.
+     */
+    void addLink(Isl isl) {
+        Switch srcSwitch = getOrInitSwitch(isl.getSrcSwitch());
+        Switch dstSwitch = getOrInitSwitch(isl.getDestSwitch());
+
+        // Remove if the same already exists.
+        srcSwitch.getOutgoingLinks().removeIf(link -> link.getDestSwitch().equals(isl.getDestSwitch()));
+        dstSwitch.getIncomingLinks().removeIf(link -> link.getSrcSwitch().equals(isl.getSrcSwitch()));
+
+        // Clone the object to unbind the src and dest fields in Isl entity.
+        Isl copiedIsl = isl.toBuilder().srcSwitch(srcSwitch).destSwitch(dstSwitch).build();
+        srcSwitch.getOutgoingLinks().add(copiedIsl);
+        dstSwitch.getIncomingLinks().add(copiedIsl);
+    }
+
+    private Switch getOrInitSwitch(Switch sw) {
+        // Clone the object to unbind the relations' fields in Switch entity.
+        return switches.computeIfAbsent(sw.getSwitchId(),
+                swId -> sw.toBuilder().outgoingLinks(new ArrayList<>()).incomingLinks(new ArrayList<>()).build());
+    }
+
+    /**
+     * Sets the equal cost for forward and reverse links between the same src and dst switches. The algorithm runs on
+     * the Outgoing and the Incoming links, which could have different values.
+     */
+    void setEqualCostForPairedLinks() {
```

Comment:
Is this method is refactored [once](https://github.com/telstra/open-kilda/blob/develop/services/src/pce/src/main/java/org/openkilda/pce/model/AvailableNetwork.java#L150)? If so, it has completely different propose.

Reply:
The purpose of the method remains the same: set the lowest cost for forward and reverse links between the same src and dst switches.
In the previous implementation, they performed "reduction" of links to leave only one with minimal cost. But with the new approach we keep both ISLs, but set equal cost.

Reply:
If I am not mistaken, initially this method performs reduction on the unidirectional links. It handles case when two switches has several physical connections between

Reply:
You're right. The reduceByCost method reduces same direction links with the same endpoints to only one. I've reverted the implementation of reduceByCost to perform this function.